src/Microsoft.Identity.Client/Cache/CacheSessionManager.cs

```diff
@@ -2,60 +2,131 @@
 // Licensed under the MIT License.
 
 using System;
+using System.Collections.Generic;
 using System.Threading.Tasks;
 using Microsoft.Identity.Client.Cache.Items;
 using Microsoft.Identity.Client.Cache.Keys;
 using Microsoft.Identity.Client.Internal.Requests;
 using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.TelemetryCore;
+using Microsoft.Identity.Client.TelemetryCore.Internal;
+using Microsoft.Identity.Client.TelemetryCore.Internal.Events;
 
 namespace Microsoft.Identity.Client.Cache
 {
+    /// <summary>
+    /// MSAL should only interact with the cache though this object. It is reponsible for firing cache notifications.
+    /// Flows should only perform (at most) 2 cache accesses: one to read data and one to write tokens. Reading data multiple times 
+    /// (e.g. read all ATs, read all RTs) should not refresh the cache from disk because of perf impact.
+    /// Write operations are still the responsability of TokenCache (TODO)
+    /// </summary>
```

Comment:
This pre-existing object was very thin and now it will take on the responsability of refreshing the cache from disk. It gets created once per request, i.e. whenever ExecuteAsync is called.